new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CleanupUtil.java

```diff
@@ -0,0 +1,524 @@
+import io.kubernetes.client.openapi.models.V1Job;
+import io.kubernetes.client.openapi.models.V1PersistentVolume;
+import io.kubernetes.client.openapi.models.V1PersistentVolumeClaim;
+import io.kubernetes.client.openapi.models.V1ReplicaSet;
+import io.kubernetes.client.openapi.models.V1Role;
+import io.kubernetes.client.openapi.models.V1RoleBinding;
+import io.kubernetes.client.openapi.models.V1Secret;
+import io.kubernetes.client.openapi.models.V1Service;
+import io.kubernetes.client.openapi.models.V1ServiceAccount;
+import oracle.weblogic.domain.Domain;
+import oracle.weblogic.kubernetes.TestConstants;
+import oracle.weblogic.kubernetes.actions.TestActions;
+import oracle.weblogic.kubernetes.actions.impl.primitive.HelmParams;
+import oracle.weblogic.kubernetes.actions.impl.primitive.Kubernetes;
+import org.awaitility.core.ConditionFactory;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static oracle.weblogic.kubernetes.extensions.LoggedTest.logger;
+import static org.awaitility.Awaitility.with;
+
+/**
+ * CleanupUtil is used for cleaning up all the Kubernetes artifacts left behind by the integration tests.
+ *
+ */
+public class CleanupUtil {
+
+  /**
+   * Cleanup all artifacts in the Kubernetes cluster.
+   *
+   * <p>Tries to gracefully delete the WebLogic domains and WebLogic Operator in the namespaces, if found.
+   * Then deletes everything in the namespaces.
+   *
+   * <p>Waits for the deletion task to be completed until up to 3 minutes.
+   *
+   * @param namespaces list of namespaces
+   */
+  public static void cleanup(List<String> namespaces) {
+    try {
+      // If namespace list is empty or null return
+      if (namespaces == null || namespaces.isEmpty()) {
+        return;
+      }
+      // iterate through the namespaces and delete domain as a
+      // first entity if its not operator namespace
+      for (var namespace : namespaces) {
+        if (!isOperatorNamespace(namespace)) {
+          deleteDomains(namespace);
+        }
+      }
+      // iterate through the namespaces and delete operator if
+      // its operator namespace.
+      for (var namespace : namespaces) {
+        if (isOperatorNamespace(namespace)) {
+          uninstallOperator(namespace);
+        }
+      }
+
+      // Delete artifacts in namespace used by the test class
+      for (var namespace : namespaces) {
+        deleteNamespacedArtifacts(namespace);
+      }
+
+      Thread.sleep(30 * 1000); // Using Thread.sleep for a one time 30 sec sleep.
+      // If pollDelay is set to 30 seconds below, its going to sleep 30 seconds for every namespace.
+
+      // wait for the artifacts to be deleted, waiting for a maximum of 3 minutes
+      ConditionFactory withStandardRetryPolicy = with().pollDelay(0, SECONDS)
+          .and().with().pollInterval(10, SECONDS)
+          .atMost(3, MINUTES).await();
+
+      for (var namespace : namespaces) {
```

Comment:
This is still synchronous by namespace, correct?  It will be curious to see how long clean-up takes.  Perhaps because clean-up happens for each test suite as that suite finishes, maybe this is parallel enough.

Reply:
Yes this is synchronous. But the delete operation on the artifacts in all namespaces(from single test class) is already called (line # 74) hopefully further iteration of the namespace won't wait long.

Reply:
i think it is worth delaying at least 5 seconds to give the apiserver a chance to process the requests...

Reply:
and i am not sure if 3 mins in too long or too short - need to work out exactly what it would be deleting at this point